A3C LSTM GA for language grounding

beduffy opened the pull request:
- Added A3C_LSTM_GA model from DeepRL-Grounding by Devendra Chaplot
- Contains cozmo_env PR
- Added checkpointing and experiment IDs of models
- Added tensorboardX which outputs to experiment folder with config+argparse settings 
- Added avg_episode_return
- Added random actions on init 
- Added 4 natural language tasks with variants specified by config files (NaturalLanguageLookAtObjectTask, NaturalLanguageNavigateToObjectTask, NaturalLanguagePickUpObjectTask, NaturalLanguagePickUpMultipleObjectTask)
- Added test case for 2 natural language tasks
- Added if statements to decide which model to use depending on whether task has natural language or not and therefore chooses A3C or A3C_LSTM_GA
- Added 3 config files and renamed main config to default_config.json
- Created task_utils.py for visualisation, preprocessing and reward functions
- Fixed config overwrite bug
- Fixed many other bugs I've forgotten about (e.g. one test case was broken)
- Renamed to PickUpTask from PickUp
- Added ability to remove lookupdown_actions and moveupdown_actions and put action alone (for pickup tasks but not put needed)
- Added ai2thor_examples.py for previous test cases that were purely just ai2thor run examples
- Added inbuilt_interactive_mode.py for script for exploring ai2thor with unity build path and `.interact()`
- Added pdb_interactive_and_check_bbox_reward_functions.py for testing reward functions and printing bboxes using pdb debugging
- VizDoom integration as originally with DeepRL-Grounding
- random scene reset choice and setting from config
- ASCII art figures for architectures within model.py
- Documentation

beduffy committed:
Added bounding_boxes_and_test_reward.py which allows experimentation with bounding boxes and visualising them and checking if we are close and focusing on an object

Added sample event metadata json for easy viewing

beduffy committed:
Added microwave or mug navigation task to wrapper

Added scene id and current object type params. Cleared up some stuff. Fixed bug in total reward.

beduffy committed:
Fixed bounding boxes to use instance_detections2D attribute instead of colorBounds

This avoided "the colored mug changing color as you place it and it not being in color_to_object_id matrix" bug. Put everything into main

beduffy committed:
Fixed "class and object info not showing in metadata on reset of env"

Added interaction param to env which removes actions which use objects. Removed some code.

beduffy committed:
Started monitoring episode lengths (3rd figure) and time taken for microwave task

Saved files by episode number rather than steps.

beduffy committed:
Added dense reward option for microwave task

TODO add more general dense reward for groups of objects

beduffy committed:
Added Devendra Chaplot's DeepRL grounding A3C model

beduffy committed:
Added natural language option to environment with 1 task

128 resolution

beduffy committed:
Added many changes! added todos, added training of A3C_GA version

Fixed 2 bugs (time embedding index failure and action output) and allowed colour channels and higher resolution. Grayscale option to environment
renamed a3c_lst_ga model file. Had to make lots of things a double though.
Stopped creating environment in main file
Renamed main file

beduffy committed:
Put all plots into utils common function and used for both models.

Created environment guid and saved plots into folder identified by that.
Added todos. Changed A3C to use 128x128 grayscale

beduffy committed:
Added checkpointing and resuming in experiment folder

Made asynchronous train the same.

beduffy committed:
Added script to read latest 3 plots saved in png

Don't spend your life looking at numbers. Automate your experiments

beduffy committed:
Added entity features into state information. 27 features. Created new model to deal with it.

For 3 mugs: top_left_screen_x, top_left_screen_y, width, height, x, y, z, 3d distance, class.
Removed legend in one lplot

beduffy committed:
Added two sentence tasks (turn left or turn right 3 times)

Append all last actions to list. Added done break condition to wrapper random walk
Made bounding crosshair function one line. Added todos

beduffy committed:
Removed reward clamp Saved action probabilities plot every episode

beduffy committed:
Added go and look and microwave and cup tasks and test case to check

task == 6 and fixed up crosshair check (centre crosshair has to be in
bounding box centre for object.
Added success return value to avoid calculating done twice (can still do
 better refactoring)
Show coloured preprocess to 128 resolution (still works in 0-1 range)

beduffy committed:
Fixed massive bug of not loading new sentence and much more
p_losses and v_losses print of average of these. Two new charts made showing these
print last reward on done
interaction=False
MOVEMENT_REWARD_DISCOUNT up one order.

beduffy committed:
Added single word instruction sentences and fixed a bug with current object type

Added better test of task to get microwave or cup and 2 embeddings.
Unsqueeze and expand in model (will check in future if is right with sentences)
many todos
CUDA device to 0 but no effect of course.

beduffy committed:
Added check for empty arrays for plotting newer graphs

Stopped rendering depth and class to get 5-10 or so FPS faster

beduffy committed:
Tried LR 0.0003, Added experiments text file, saved episode number
Fixed bug with checking to checkpoint at every step

Many todos

beduffy committed:
Added checkpoint and first experiment folder. Fixed latest plot blug.

Checkpoint every 100k

beduffy committed:
Added multi-threading, save total length/number of episodes, counter, timing and benchmarking

print_all and print for rank = 0.
some todos, commented out eids

beduffy committed:
Fixed checkpoint_counter for new experiment, some todos and comments

beduffy committed:
Added .gitignore for experiments

fernandotorch committed:
Add Rainbow DQN code

Adapted from RL-adventure repo. First running version, although results look odd.

beduffy committed:
Some comments, style and prognostications

fernandotorch committed:
Merge branch 'A3C_barebones' into Rainbow-DQN

fernandotorch committed:
Fix channel dimension from state

There was a bug on the channel order returned from the environment by the function step. 
Rainbow dqn code also adapted to use the fix

beduffy committed:
Merge branch 'A3C_LSTM_GA_two_sentence' into A3C_LSTM_GA_barebones
Fixed conflicts and 16 files.

beduffy committed:
Added A3C GA model to model.py and removed 3 files I won't need (good-experiment-ids.txt, model.py, check_latest_plots)

Moved sample event metadata to gym_ai2thor folder

beduffy committed:
Import A3C_LSTM, added natural language param, some copying into train.py file but commented natural language (unfinished), beginnings of natural language task

comment fix square input and README a3c better english

beduffy committed:
Added cameraY, gridSize, incremental_rotation_mode with 10 degree rotations params. Loading build from unity. Rotation broke

beduffy committed:
Added new task example for agent on the ground with cups. Added cameraY, gridSize, rotation, build_path to config_example.json. set reasonable defaults and checks

beduffy committed:
Added continuous mode to initialise of ai2thor which finally allowed moving in any direction

beduffy committed:
Added argparse to continuous example for build path, default false incremental rotation, fixed 1 test by gridSize 0.25, both to readme config.json example

beduffy committed:
NLP instructions working. Added natural_language if/else statements, cleaned up NaturalLanguageLookAtTask and NL environment settings, removed batch forgotten if statements, a3c_ga to float, many new functions in utils and elsewhere

beduffy committed:
Merge branch 'master' of https://github.com/TheMTank/ai2thor-experiments into A3C_LSTM_GA_barebones

beduffy committed:
Renamed to Bowl and Mug, passed curr_object_type in event to task, expand_as, euclidean distance for close_enough

comments and documentation

beduffy committed:
Added uuid eid args and old checkpointing, todos, references to old commented ways to delete in next commit, timing, todos, printing of policy/value loss

beduffy committed:
save_checkpoint function with os paths (called in args num_steps), cleared commented checkpoint code and tested, moved bounding_boxes code to tests

beduffy committed:
Deleted 8 old files. Old wrapper finally gone (entity feats within if needed). Old A3C thread stats gone in main.py. Old train.py with some nice comments on how to split up and standardise common RL algorithms. Old train_a3c_lstm_ga.py and utils.py

beduffy committed:
Added option to disable lookupdown actions, added option to do num random actions on reset, fixed utils read config bug and saved latest argparse params to json file in experiment folder

beduffy committed:
Added NaturalLanguageLookAtObjectTask, NaturalLanguageNavigateToObjectTask, nlpbase task, removed all natural language from Env with modified step function with extra state info from task, task_utils.py for preprocessing and reward functions, tensorboard import, load and finally export

beduffy committed:
Added a new large test case for naturalLanguageLookAtTask which tests that 5 different objects work and added env.close() to end of many test cases

beduffy committed:
Added two examples, inbuilt_interactive_mode.py with argparse for unity build and 2nd was renamed bounding_boxes file. Removed old functions in file and use gym_task_utils. Added line option from bounding boxes to crosshair and printed distance

beduffy committed:
Removed natural_language_instructions from config since it's decided by task, check if reset before first step and raise ValueError otherwise, random seed

beduffy committed:
Moved show_bounding_boxes() to task_utils.py with extra features, raise ValueError if objectType not found, pass 2d+3d distance thresholds to lookAt reward functions. Added verbose option for debugging. default reward is 1 and list of instructions can be passed

beduffy committed:
Fixed merge conflicts from merging cozmo_env. ai2thor_env.py

beduffy committed:
Made sure grayscale on/off worked, added args.task_name, removed all mentions of args.natural_language, added todos, turn_instruction_str_to_tensor to task_utils.py, apple and mug default, controller.stop() at end of other test cases without wrapper.

Added warning in comment about bad build path.

beduffy committed:
Added writer to train and saved scalars into tensorboard, added cozmo env params (gridsize, camerayY, buildpath), added and removed todos, max_episode_length to config_example.json, pass whole config to task, extra README

beduffy committed:
Fixed max_episode_length bug and passed kwargs correctly all the way to BaseTask, renamed string to PickUpTask, renamed file without on_ground, gridSize 0.05,

beduffy committed:
Added NaturalLanguagePickUpObjectTask, Added movement_reward, incremental_rotation, saved all_scalars to experiment folder, saved episode_number and episde length correctly in tensorboard

beduffy committed:
Added 3 extra config files (for specific task variations) and renamed to default_config.json

beduffy committed:
Fixed test.py for both A3C+A3C_GA, added negative reward for picking up wrong object and added comments to pdb_interactive

beduffy committed:
Fixed reset bug by setting step_num to 0, changed to build_file_name in build_files folder, added moveupdown_actions option

beduffy committed:
Added calculate_lstm_input_size_for_A3C_LSTM_GA with hardcoded factors and used it, docstring to A3C_LSTM_GA and inputted vocab_size, episode_length. Added todos, comments

beduffy committed:
Changed task random walk example to take build_file_name and auto-calculated x_attention dimensions by changing calculate_input_width_height func to return tuple

beduffy committed:
Fixed off by one episode_length time embedding by incrementing after step, moved 3 ai2thor tests to ai2thor_examples.py, default_config PickUpTask and movement_reward in others,

beduffy committed:
Added test case for NaturalLanguagePickUpObjectTask, removed 3 ai2thor test cases and made them examples, added config file name and checkpoint-freq arg params, saved latest_config.json and removed large config_dict in main.py. Initialised arg.episode_number and total_length outside of params.

printed current object picked up. added todos,

beduffy committed:
Renamed incremental_rotation to continuous_movement, added a few comments, startswith+endswith for Object interactions

beduffy committed:
Added todos, printed elapsed training time and time for 20 steps only every 100 backprops, distance_threshold_3d to task config for nl_lookat task

beduffy committed:
Added num-random-actions-at-init to argparse, max_episode_length back, added comments, commented word_embedding expand (must understand), added todos, saved image correctly, print predicted value for now to find out why spikes,

beduffy committed:
Added verbose-num-steps arg, printed tensorboard command, shutil import and model_is_best param, added avg_rewards with avg_over_num_episodes, fixed image tensorboard bug and reordered some code

beduffy committed:
Cleaned comments, removed some todos, changed to floorplan2 in rotate only, absolute rotation to 0.0 on reset,

beduffy committed:
Cleaned+added comments and removed some todos

beduffy committed:
Fixed multi-word sentence bug, rotate_only will have multi-word sentences since otherwise too easy, removed final todos in model.py

beduffy committed:
Removed todos regarding predicted value spike (didn't find reason), removed rainbow files so PR looks neater

beduffy committed:
Fix merge conflicts from merging cozmo_env into A3C_GA. continuous_movement instead of incremental

beduffy committed:
Fixed test() function, renamed file, added many comments, fixed NL_pickup task and turned off lookupdown_actions, if open_close check for allowed_objects, allow to specify pickup_crosshair_distance in task requiring pickup

 Only writing to tensorboard if rank == 0,

beduffy committed:
Renamed avg_rewards to avg_episode_returns, added euclidean distance check to pickup because it was too large as default, created calculate_euc_distance_between_agent_and_object() function,

beduffy committed:
Added 4 vizdoom_data files, 1 vizdoom_maps room.wad file, 3 python vizdoom_utils files and added GroundingEnv into envs.py with comment splitting between Atari

beduffy committed:
Renamed envs to env_atari.py and took out vizdoom part into env_vizdoom.py. Movement reward set to -0.001 and pickup euclidean distance to 0.3

beduffy committed:
Got VizDoom working in main.py and model.p.y Added ViZDoom argparams, better experiment-id with date+uuid if not specified, inputted resolution to model and calculated non-square lstm_cell_size and ordered correctly for vizdoom, DeepRL-Grounding multiprocessing error fix

Added better docstring to env_vizdoom and todo

beduffy committed:
Added random scenes on reset option within task, build_bowls_vs_cups_fp1_v_0.1 config added (first versioned build), ai2thor_inbuilt_interactive_mode broken and todos

beduffy committed:
Added a3c/main.py variants to README.md with more info on natural language tasks. Added seed() with warning saying not implemented to GroundingEnv so it does not crash, renamed max_object_pickup vars

beduffy committed:
Got train and test working in vizdoom, added beautiful unpack_state() function to task_utils.py and used 3 times in train, renamed to image_state, passed resolution to models and added 2d euclidean distance option for function

beduffy committed:
Got VizDoom running from any folder by converting relative paths to absolute paths for data, changed vizdoom reset() to only return 1 thing from 4, seed function fix

beduffy committed:
Fixed glob for latest checkpoint, fixed ai2thor inbuilt interactive by clearly saying terminal only, check if unity build file path exists in two places

beduffy committed:
Added complicated logic for args.resume-latest-config (default is no override allowed from config file), reordered main.py and more comments, Fixed missing R problem, added 4 room config file, fixed euclidean+crosshair distance defaults. Print new scene_id on change

Added todo to fix multiprocessing
print Exception as e around process creation
removed --dump-location and early attempts at logging.info

beduffy committed:
Removed args.task-name, removed Exception as e, tensorboard creation to bottom, removed microwave or cup checkpoint, added todo to utils and a bit more info to README.md about config+task variants

beduffy committed:
Added NaturalLanguagePickUpMultipleObjectTask with no terminal with config file, Added ability to remove put_interaction alone, allowed multiple objects in inventory in wrapper and in build file C# code, printed full inventory object types before and after

put_interaction = false on all NL pickup tasks

beduffy committed:
Added num_backprops to checkpoint and reload, checked for "and not" pickup_interaction as well, only add scalar for value+policy loss for rank == 0 in train

beduffy committed:
Added many comments to A3C train code to aid understanding, reshuffled things and possibly fixed big bug on not calculating final reward in total reward for episode!?!. Renamed R to curr_return and understood predicted value as current return if not terminal. Reordered main.py argparse with comment and todo regarding tau rename and being bigger than gamma?

beduffy committed:
Added warning for empty acceptable_receptacles config param and fixed if statement checks in 3 places regarding pickup put interaction. Added NL_pickup multiple bowls_vs_cups and 5 objects configs+task variants. Set default config file in main to default_config.json and gave advice on other configs in help.

made eid more prominent and added todo regarding README.md

fernandotorch commented on .gitignore:
```diff
@@ -1,4 +1,4 @@
 \.idea/
 
 *.pyc
+/experiments/*
```
Shouldn't every experiment store its results in a folder within it? For rainbow I am automatically creating two folders the first time you run an experiment, "results" for plots and tensorboard summaries (2nd I don't use) and "weights" to store trained models.
You can add to ignore the following if we agree on that:
```suggestion
/*/experiments/*
```

beduffy replied:
Within the folder of
/experiments/5_objects_pickup_fp211_easy
I have 
/experiments/5_objects_pickup_fp211_easy/checkpoints
/experiments/5_objects_pickup_fp211_easy/tensorboard_logs
/experiments/5_objects_pickup_fp211_easy/latest_args.json
/experiments/5_objects_pickup_fp211_easy/latest_config.json

So not fully sure what you mean? I can run thousands of experiments, are you saying you can only run 1 with Rainbow?

fernandotorch replied:
For now yes I always override the old one, but that is completely out of the point here. What I am suggesting is that since algorithms are self-contained messiness they should also contain their own experiments in a folder a3c/experiments, rainbow/experiments and so on. Because the results you are storing might be even totally different as it was the case until now. Let me know what you think

beduffy replied:
yeah, good point, i'll do this. Also necessary since PPO could be using visdom and different folder structures. We can copy and paste the experiment folder creation code (might not be DRY but it'll be ok)

fernandotorch commented on README.md:
```diff
@@ -137,7 +149,7 @@ microwave.
 ## The Team
 
-[The M Tank](http://www.themtank.org/) is a non-partisan organisation that works solely to recognise the multifaceted 
+[MTank](http://www.themtank.org/) is a non-partisan organisation that works solely to recognise the multifaceted 
```
Not an issue for you here but in general. We should change the message from MTank to something more related to what we do now. This seems very outdated

beduffy replied:
yep, i can do it here if we think of something

fernandotorch replied:
opened issue in slack

fernandotorch commented on algorithms/a3c/env_atari.py:
```diff
@@ -4,11 +4,20 @@
 This contains auxiliary wrappers for the atari openAI gym environment e.g. proper resizing of the
 input frame and a running average normalisation of said frame after resizing
 """
-import cv2
```
cv2 is used in _process_frame42 function. Either get back the import or delete the function

beduffy replied:
thanks!

fernandotorch commented on algorithms/a3c/env_atari.py:
```diff
@@ -35,7 +44,7 @@ def _process_frame42(frame):
 class AtariRescale42x42(gym.ObservationWrapper):
```
Shouldn't def _process_frame42 be a staticmethod of this class? you are only using this method for this particular class

beduffy replied:
well that's a conversation to have regarding something that I noticed you doing before. For A3C I made the calculate_linear_size() functions as functions but you made your version in Rainbow a static class. What's your reasoning?

Why do a static class when you can do an outside function?

fernandotorch replied:
By putting it in the class you are explicitly "saying" that it is used mainly (if not only) for this class in particular. Making an outside function makes it look like you want to reuse it somewhere else which is not the case here. 
A @staticmethod is a method that you can call without needing an instance of the class which in this case I don't know if it applies but could be e.g. to calculate the size of the linear layers without having to build the model (to parametrize some post processing or things like that). Don't you think it belongs inside the class?

beduffy replied:
Some things are neater as util helper functions outside the class. E.g. for PyTorch modules I like to keep them simple e.g. init and forward and maybe 1-2 other functions. Calculating linear size is borderline outside inside.

But I do get your point. Hard to say

fernandotorch commented on algorithms/a3c/env_atari.py:
```diff
@@ -35,7 +44,7 @@ def _process_frame42(frame):
+        self.observation_space = spaces.Box(0.0, 1.0, [1, 42, 42])
 
     def _observation(self, observation):
         return _process_frame42(observation)
```
what is the whole unbaised mean thing about? Should we add some docstrings or can you explain me?

fernandotorch commented on requirements.txt:
```diff
@@ -1,2 +1,4 @@
\ No newline at end of file
+scikit-image>=0.14.1
+pytorch>=0.4.1
+
```
Add also 
```suggestion
vizdoom # needed for vizdoom experiments
```
or similar

beduffy replied:
but we don't want to discourage anyone from using our repo. ViZDoom is extremely hard to install and has so many dependencies. Our repo should be easy to start for 95% of settings.
The second you try vizdoom it will crash and they will know what to install
So my solution is if comments work in .txt files
`#vizdoom # needed for vizdoom experiments`

fernandotorch replied:
vizdoom is as hard to install as `sudo tons of ubuntu packages`, then `pip install vizdoom`. For me it took 2 minutes in windows (not supported by us) and ubuntu 16 and 18.
If you don't want to add it to requirements then add a comment in the readme to "install vizdoom if you want to run xxx experiments. Installations can be found here (link).

beduffy replied:
you are one of the lucky ones

but added :

(you will need to install ViZDoom) in README.md in the vizdoom example command

fernandotorch commented on algorithms/a3c/env_vizdoom.py:
```diff
@@ -0,0 +1,379 @@
+
+    def game_init(self):
+        """Starts the doom game engine."""
+        game = DoomGame()
```
```suggestion
        import vizdoom?
        game = vizdoom.DoomGame()
```

beduffy replied:
I have a todo at the top of this file:

`from algorithms.a3c.vizdoom_utils.doom import *   # todo probably avoid this but at least it is self-contained within this file`

importing everything is a sin/very bad practice. I could change it to import each specific thing (everything that goes red after I remove it. I suppose you agree.

fernandotorch replied:
Yes, please do!

beduffy replied:
I did this. Let me know if you disagree
![image](https://user-images.githubusercontent.com/9202993/54498154-1eaa9e00-48fb-11e9-9b0f-9021dbedac4a.png)

fernandotorch commented on algorithms/a3c/env_vizdoom.py:
```diff
@@ -0,0 +1,379 @@
+    def __init__(self, args):
+        """Initializes the environment.
+        Args:
+          args: dictionary of parameters.
```
Please explain at least how the instructions json needs to be formatted. They are probably just words separated by spaces but tell me

fernandotorch commented on algorithms/a3c/env_vizdoom.py:
```diff
@@ -0,0 +1,379 @@
+AgentLocation = collections.namedtuple("AgentLocation", ["x", "y", "theta"])
+
+
+class Task:
```
It would be much cleaner to inherit from BaseTask from our tasks.py. You would only need to add a call to super and change the name of the function "get_reward" to make it happen.

beduffy replied:
It might be a bit risky since this is a different meaning of task and I purely wanted to just make it work with ViZDoom and not focus too long on making ViZDoom and our task interface similar. 

In other words, the functionality of the true task of ViZDoom environment is within GroundingEnv, not that fake task class just so I can have that variable. Feel free to argue.

fernandotorch replied:
Yeah but since it is easy to abstract and then makes users' life easier, I'd say we do it. You just have to put the get_reward in a task and few little things more. I think it is totally worth it

beduffy replied:
Do you still think I should do this? Moving vizdoom specific code into ai2thor tasks.py is risky and hard to do for a few reasons. Why try to make everything fit into our tasks structure, especially since I didn't write this code?

fernandotorch commented on algorithms/a3c/env_vizdoom.py:
```diff
@@ -0,0 +1,379 @@
+        instr_contains_color = False
+        # instr_contains_color is set if the instruction contains the color
+        # attribute (e.g.) "Go to the largest green object".
+        # instr_contains_color is True if the instruction doesn't contain the
```
What is this madness? instr_contains_color == True only if there is NO color instruction?

fernandotorch replied:
Ok docstring error, the function looks good.
```suggestion
        # instr_contains_color is True if the instruction contains the
        # color attribute. (e.g.) "Go to the smallest green object"
```

fernandotorch replied:
Also I know that you have fobia to docstrings with """ instead of #, but I would say the rule of thumb is: If your comments are 3 lines or bigger then """, else #. Moreover, since the comment happens just after the docstring, why don't you add this info to the docstring itself?

beduffy replied:
ok

fernandotorch commented on algorithms/a3c/env_vizdoom.py:
```diff
@@ -0,0 +1,379 @@
+        self.instruction, instruction_id = self.get_random_instruction()
+
+        # Retrieve the possible correct objects for the instruction.
+        correct_objects = self.get_target_objects(instruction_id)
```
all this "correct object" sounds strange. Could you rename it and similar variables as "target object"? There is no correct or incorrect, but object selected as target/goal and others that are not

beduffy replied:
object is correct if you pick up the correct object. But meh, i don't think I should rename too much of devendra's code

fernandotorch commented on algorithms/a3c/env_vizdoom.py:
```diff
@@ -0,0 +1,379 @@
+        return reward
+
+    def get_agent_and_object_positions(self):
+        """Get agent and object positions based on the difficulty
```
Docstrings should look like this 
```suggestion
        """
        Get agent and object positions based on the difficulty
```
Stop being so stingy with vertical space

fernandotorch replied:
Also extend the docstring to explain more about the easy, medium and hard difficulties

fernandotorch commented on algorithms/a3c/env_vizdoom.py:
```diff
@@ -0,0 +1,379 @@
+
+        if self.params.difficulty == 'easy':
+            # Agent location fixed in Easy.
+            agent_x_coordinate = 128
```
If you love so much vertical space, you can put agent x, y and orientation in one line
```suggestion
            agent_x_coordinate, agent_y_coordinate, agent_orientation = 128, 512, 0
```

fernandotorch replied:
since agent_x, y and rotation are the same for easy and medium I would write it outside the if elses to avoid repeating code 
```suggestion
            agent_x_coordinate = 128
    # Then this
          if self.params.difficulty == 'easy':
               # only code to generate x and y.
          if self.params.difficulty == 'medium':
               # only code to generate x and y.
          if self.params.difficulty == 'hard':
               # override x, y and orientation defaults
```
Also one thing, is it possible to have difficulty easy and medium? If not use if elif else instead of several ifs

beduffy replied:
Ok I'll put in one line 

Explicit is good sometimes and in the future it might be good to change it specifically for medium. But fine i'll initialise it outside.

But im really agonizing about being careful about changing this code without testing it. Sometimes style and other changes can break it no matter how much you reason about the output. This is why we shouldn't really change much of this code. BUT FINE

fernandotorch commented on algorithms/a3c/vizdoom_utils/points.py:
```diff
@@ -0,0 +1,217 @@
+        rand = (random.uniform(0, length), random.uniform(0, width))
+
+    for i in range(100):
+        data = grid.poisson(rand, n_points)
```
why poisson? explain

beduffy replied:
the docstring at the top of the file gives more details 
`Poisson-disc algorithm produces points in a grid,
but no closer to each other than a specified minimum distance`
and links to more details

fernandotorch replied:
You are right. Closing issue

fernandotorch commented on algorithms/a3c/env_vizdoom.py:
```diff
@@ -0,0 +1,379 @@
+        self.correct_location = np.random.randint(5)
+
+        # Randomly select one correct object from the
+        # list of possible correct objects for the instruction.
```
So you select one object randomly if there are two corrects? This is confusing. If you tell me go to green pillar and there are 2 I should get my reward and be happy regardless of the one I choose shouldn't I?

beduffy replied:
maybe this case is very rare. Do you want me to go into it?

fernandotorch commented on algorithms/a3c/env_vizdoom.py:
```diff
@@ -0,0 +1,379 @@
+                         chosen_correct_object)
+            object_ids = [self.object_dict[x] for x in object_ids]
+
+        assert len(object_ids) == 4
```
assert shouldn't be here. Either check and raise an exception or ignore it always.

beduffy replied:
Just a question to you, are you saying checks and raising an exception is always preferred to an assert?
The first could be 2-3 lines whereas an assert is often self-explanatory, neater, more concise and 1 line

fernandotorch replied:
Assert is for testing, I can ignore them whenever I want. To ignore an exception you need to change the code.
IMO the place for assertions is tests, but feel free to argue.

beduffy replied:
Call Python with the -O flag to avoid asserts apparently. But why would someone do this with our file specifically?

asserts are concise. Raising Exceptions are ugly.

it might be fine to swap here but not in a3c class

fernandotorch commented on algorithms/a3c/env_vizdoom.py:
```diff
@@ -0,0 +1,379 @@
+
+        # Special code to handle 'largest' and 'smallest' since we need to
+        # compute sizes for those particular instructions.
+        if 'largest' not in self.instruction \
```
Ifs should be defined by positives if possible. 
```suggestion
        if 'largest' in self.instruction ...:
            # code currently in else
        else:
           # code currently in if
```

beduffy replied:
I've never heard that but I've always felt that. Did PEP8 or something else tell you this?

fernandotorch replied:
PEP8 https://www.python.org/dev/peps/pep-0008/#programming-recommendations
Use is not operator rather than not ... is. While both expressions are functionally identical, the former is more readable and preferred.
```
Yes:

if foo is not None:
No:

if not foo is None:```

fernandotorch replied:
you say it as if pep8 was something bad. I think this is a good pattern in general

beduffy replied:
we have to be so careful that this doesn't ruin the logic

beduffy replied:
yeah omg I did this but that is completely different logic just because you had a style change!!!!!! 

These two code samples are not equivalent

```
        if 'largest' not in self.instruction and 'smallest' not in self.instruction:
            object_ids = random.sample([x for x in range(len(self.objects))
                                        if x not in correct_objects], 4)
        else:
            object_ids = self.get_candidate_objects_superlative_instr(
                         chosen_correct_object)
            object_ids = [self.object_dict[x] for x in object_ids]
```

```
        if 'largest' in self.instruction and 'smallest' in self.instruction:
            object_ids = self.get_candidate_objects_superlative_instr(
                chosen_correct_object)
            object_ids = [self.object_dict[x] for x in object_ids]
        else:
            object_ids = random.sample([x for x in range(len(self.objects))
                                        if x not in correct_objects], 4)
```
you can't have largest and and smallest in the same instruction but of course you can check that both are not in it. Inversion of logic was ruined

fernandotorch commented on algorithms/a3c/env_vizdoom.py:
```diff
@@ -0,0 +1,379 @@
+            self.game.make_action(actions[action_id], 5)
+
+        self.time += 1
+        reward = self.get_reward()
```
For consistency use
```suggestion
        reward, done = self.task.transition_reward(self.event)
```

fernandotorch replied:
Using different function names and conventions at every algorithm makes it harder for a newcomer to "get" what we are doing.

beduffy replied:
I didn't write this ViZDoom code, so I'm not gonna change their env class to be like ai2thor env class, we just wanted it to work, not to rewrite it.

fernandotorch replied:
It is not too much effort to modify that, but if you prefer it ugly then leave it as is.

beduffy replied:
you're suggesting to put vizdoom code into our tasks.py. that's a big ask

beduffy replied:
if you agree that we should use ViZDoom because of these reasons: 
1. We already have it installed
2. It's super customisable (check these tools)
3. Up to 7000 fps in sync mode, single threaded

Then I'll try to make our tasks.py work with it

fernandotorch commented on algorithms/a3c/env_vizdoom.py:
```diff
@@ -0,0 +1,379 @@
+                               target_location.x, target_location.y)
+        if dist <= REWARD_THRESHOLD_DISTANCE:
+            reward = CORRECT_OBJECT_REWARD
+            return reward
```
I don't like this pattern of putting returns at different stages of the code specially since here it is not needed at all

beduffy replied:
ok removed here

fernandotorch commented on algorithms/a3c/env_vizdoom.py:
```diff
@@ -0,0 +1,379 @@
+                                       object_location.x, object_location.y)
+                if dist <= REWARD_THRESHOLD_DISTANCE:
+                    reward = WRONG_OBJECT_REWARD
+                    return reward
```
Add break instead of return. Same idea as in the last comment above

beduffy replied:
`reward = self.params.living_reward` 
is the line after so this will actually break and you'll never get the wrong object reward

fernandotorch commented on algorithms/a3c/env_vizdoom.py:
```diff
@@ -0,0 +1,379 @@
+        random.shuffle(output_objects)
+        return [x.name for x in output_objects[:4]]
+
+    def get_objects_info(self):
```
Could you add docstrings to this? Like `This is used to get objects with sizes` or whatever information you intend to get with this.

fernandotorch replied:
Same for get_all_objects but not get_target_objects (this last one is clear by itself)

fernandotorch commented on algorithms/a3c/env_vizdoom.py:
```diff
@@ -0,0 +1,379 @@
+
+        return instruction, instruction_id
+
+    def get_word_to_idx(self):
```
This is very odd. If you want to have words with indices use a list instead.

beduffy replied:
I don't understand. word_to_idx and idx_to_word are the two most common NLP coding patterns.

fernandotorch replied:
If you want to get elements by index or name a list is easier to use than a dictionary. 
My suggestion is changing this `word_to_idx[word] = len(word_to_idx)` to this `word_to_idx.append(word)` (of course the object definition has to be changed as well). Isn't that obviously better? Why do you need keys and values instead? 
Since what I am telling you is the standard in the python world, could you please argue why something different is needed here?

beduffy replied:
have you seen NLP code before?

we need to go from word to idx, not idx to word. The list works only in the latter but not the former case

How else are you going to find the word embedding index from just a string?

fernandotorch commented on algorithms/a3c/main.py:
```diff
@@ -10,28 +11,41 @@
 import torch.multiprocessing as mp
+from tensorboardX import SummaryWriter
 
 from gym_ai2thor.envs.ai2thor_env import AI2ThorEnv
```
The order should be alphabetic. Algorithms first and them gym

beduffy replied:
I think ordering imports by alphabetic is going too far :P

fernandotorch replied:
Disagree with you, but fine.

fernandotorch commented on algorithms/a3c/main.py:
```diff
@@ -10,28 +11,41 @@
 parser.add_argument('--gamma', type=float, default=0.99,
-                    help='discount factor for rewards (default: 0.99)')
+                    help='discount factor for returns (default: 0.99)')
+# todo rename tau to lambda and find out why it is higher than gamma?
```
Remember to finish this

fernandotorch commented on algorithms/a3c/main.py:
```diff
@@ -40,80 +54,275 @@
 
 if __name__ == '__main__':
-    os.environ['OMP_NUM_THREADS'] = '1'
     os.environ['CUDA_VISIBLE_DEVICES'] = ""
```
why is cuda not allowed?

beduffy replied:
A3C is usually CPU only. It might require a slightly different implementation to work on GPU e.g.
https://github.com/NVlabs/GA3C

fernandotorch replied:
Can you add a comment explaining it and link to that repo you copied here? This is useful info

fernandotorch commented on algorithms/a3c/main.py:
```diff
@@ -40,80 +54,275 @@
+        args.resolution = (42, 42)  # fixed to be 42x42 for _process_frame42() in envs.py
+    elif args.vizdoom:
+        # many more dependencies required for VizDoom so therefore import here
+        from algorithms.a3c.env_vizdoom import GroundingEnv
```
Nope nope. I don't want imports somewhere else than the top

beduffy replied:
have you not seen this style and very very very correct justifications for this...
I even told you this in slack and you didn't say anything.
This is perfectly acceptable. Requiring ViZDoom should NEVER be a requirement until the second you run it @fgtoralesch

fernandotorch replied:
Fine. Let's leave it there then

fernandotorch commented on algorithms/a3c/main.py:
```diff
@@ -40,80 +54,275 @@
+        # many more dependencies required for VizDoom so therefore import here
+        from algorithms.a3c.env_vizdoom import GroundingEnv
+
+        if args.evaluate == 0:
```
I'd rather have strings than 0, 1 and 2. If we start adding this codes then people will have to memorize them to use the code. I'd do:
``` --evaluate "train" ```
And be happier than 0

beduffy replied:
I agree but i might keep it the same due to me not writing the ViZDoom part. If we ever extend the ViZDoom part we can change it all we want. Or feel free to give one more push

fernandotorch replied:
Go to slack for discussing this. In general I would say don't touch this ugly code and leave it for reproducibility of the original code. But this hurts deep inside because we are constantly adding and ignoring the fact that we have this weird things in OUR repo

fernandotorch replied:
From the second we start using vizdoom further from reproducing the original training I would refactor it. For now let's leave it as it is.

fernandotorch reviewed (changes requested):
First pass of PR review. Went only through 2 files so there is still a lot to do

beduffy committed:
Fixed self.controller.start() indent bug, added missing cv2 import to env_atari.py, fixed docstring about square image. Added comment about detach.

added todo

fernandotorch commented on algorithms/a3c/main.py:
```diff
@@ -40,80 +54,275 @@
-                    help='maximum length of an episode (default: 1000000)')
 parser.add_argument('--no-shared', default=False,
                     help='use an optimizer without shared momentum.')
+parser.add_argument('--num-processes', type=int, default=4,
```
Any info on how these processes are distributed within different cores? Or mention/link the library used so people know how it is handled

fernandotorch replied:
Couldn't we have 1 parameter instead of 2, i.e. num-processes and synchronous? From what I read here I assume that synchronous just means num-processes == 1. Or is it sth else as well?

beduffy replied:
You can have 1 process and asynchronous because it makes a new a thread and you can't debug within it then

fernandotorch commented on algorithms/a3c/main.py:
```diff
@@ -40,80 +54,275 @@
 parser.add_argument('--atari-env-name', default='PongDeterministic-v4',
                     help='environment to train on (default: PongDeterministic-v4)')
-#
 parser.set_defaults(atari=False)
```
Isn't the default already False? only if you `--atari` it will be true. Why doing it twice? same applied to atari_render

fernandotorch commented on algorithms/a3c/main.py:
```diff
@@ -40,80 +54,275 @@
+parser.add_argument('--sleep', type=float, default=0,
+                    help="""Sleep between frames for better
+                    visualization (default: 0)""")
+parser.add_argument('--vizdoom', dest='vizdoom', action='store_true',
```
What I did with rainbow was to let people select "game", so the argument is a string saying "ai2thor", "space-invaders" or any of the many environments out there. I suggest you use a similar parameter here instead of one per environment

fernandotorch commented on algorithms/a3c/main.py:
```diff
@@ -40,80 +54,275 @@
-        args.frame_dim = env.config['resolution'][-1]
-    shared_model = ActorCritic(env.observation_space.shape[0], env.action_space.n, args.frame_dim)
-    shared_model.share_memory()
+        # if you resume a checkpoint and if args.resume_latest_config == True, changes to the file
```
Use """ instead of three #

fernandotorch reviewed (commented):
Few more comments and answer to yours above

beduffy committed:
Merge branch 'master' into A3C_LSTM_GA_barebones with cups-rl name change

beduffy committed:
A3C calculate_lstm_input_size_for_A3C() doesn't need square images by using resolution[1]

beduffy committed:
Made README.md much better (split into subsections and gave contructor example), Added todo about seed in ai2thor, default reward for 5 objects is 5 and added to natural language tasks, removed warning of overwriting if the key is the same (great idea)

fernandotorch commented on algorithms/a3c/main.py:
```diff
@@ -40,80 +54,276 @@
+
     args = parser.parse_args()
+    # set to 0 so that checkpoint resume can overwrite if necessary
+    args.episode_number = 0
```
You have a parameter to load the checkpoint so use it although I couldn't find it. I suggest something on these lines
```suggestion
    args.episode_number = args.checkpoint['episode_numer'] if args.checkpoint else 0
```
Also please add a docstring explaining what should I do to load an experiment and/or modify it. Right now it is all confusing to me

beduffy replied:
I'm actually loading in the default value as 0 and then later on after reading in checkpoint (this requires a lot of logic to check if you even have a checkpoint) I overwrite

But sure more docstrings needed

fernandotorch commented on algorithms/a3c/main.py:
```diff
@@ -40,80 +54,276 @@
+    if not args.experiment_id:
+        args.experiment_id = datetime.datetime.now().strftime("%Y-%m-%d-") \
+                                                                     + str(uuid.uuid4())
+    args.experiment_path = os.path.join(os.path.abspath(os.path.join(os.path.dirname(__file__),
```
did you test if this works from any folder? I assume yes

beduffy replied:
yep, double checked

fernandotorch commented on algorithms/a3c/main.py:
```diff
@@ -40,80 +54,276 @@
 
-    torch.manual_seed(args.seed)
+    # Create environment from choices of Atari, ViZDoom and our ai2thor wrapper with tasks
     if args.atari:
```
As mentioned above, don't use 3 arguments for 3 different environments. args.environment or args.game and check the string you have there "vizdoom", "space_invaders" or whatever.

beduffy replied:
I did this. atari, vizdoom and ai2thor. the default atari is pong so you need two arguments to decide which atari env which is fine to me

fernandotorch commented on algorithms/a3c/main.py:
```diff
@@ -40,80 +54,276 @@
+        env = GroundingEnv(args)
+        env.game_init()
+        args.resolution = (args.frame_width, args.frame_height)
     else:
```
else should raise an exception for "environment not supported". Check always strings of env name even for ai2thor

fernandotorch commented on algorithms/a3c/main.py:
```diff
@@ -40,80 +54,276 @@
+        args.last_config_resume_path = os.path.join(args.experiment_path, 'latest_config.json')
+        args.config_dict = {}
+        if args.resume_latest_config:
+            print('args.resume_latest_config is set on')
```
```suggestion
            print('Resuming latest config for the environment and task')
```

fernandotorch commented on algorithms/a3c/main.py:
```diff
@@ -40,80 +54,276 @@
+                          'Set to args.resume_latest_config to 0 if you don\'t want this')
+                    args.config_dict = json.load(f)
+        else:
+            print('Loading most params from args.config-file-name')
```
Too vague. I would say something like this:
```suggestion
            print('Loading default params from args.config-file-name. Note that these will be overriden if other args were given')
```

fernandotorch commented on algorithms/a3c/main.py:
```diff
@@ -40,80 +54,276 @@
+                      .format(args.experiment_id))
+                with open(args.last_config_resume_path, 'r') as f:
+                    print('Loading config_dict from file: {}'.format(args.last_config_resume_path))
+                    print('Therefore changes to args.config_file_name won\'t affect the env.\n'
```
Rephrase it for clarity
```suggestion
                    print('Changes to args.config_file_name won\'t affect the env.\n' 
                    'Set to args.resume_latest_config to 0 to override args.config_file with params from xxx'
```

fernandotorch commented on algorithms/a3c/main.py:
```diff
@@ -40,80 +54,276 @@
+                         seed=args.seed)  # todo decide
+        args.resolution = (env.config['resolution'][0], env.config['resolution'][1])
 
+    # Create shared model
```
Add an explanation like `A3C uses a shared model that is updated with the results of rollouts in parallel blabla.` Put some context to what "shared model" means

fernandotorch commented on algorithms/a3c/main.py:
```diff
@@ -40,80 +54,276 @@
-    lock = mp.Lock()
+    # Checkpoint creation/loading below
+    checkpoint_counter = False
+    if not os.path.exists(args.checkpoint_path):
```
pep8: change if for a positive condition, i.e. if os.path.exists(bla) do what know is in the else, else do the things that are here

beduffy replied:
i found it easier to think about:
if no checkpoint folder -> create checkpoint folder
else -> DO LOTS OF SHIT TO READ THE CHECKPOINT FOLDER.

just easier to think about no?

with the other way you have to scroll a bit to find out what is happening

fernandotorch commented on algorithms/a3c/main.py:
```diff
@@ -40,80 +54,276 @@
+    # Checkpoint creation/loading below
+    checkpoint_counter = False
+    if not os.path.exists(args.checkpoint_path):
+        print('Tensorboard created experiment folder: {} and checkpoint folder '
```
```suggestion
        print('Creating tensorboard experiment folder: {} and checkpoint folder '
```

fernandotorch replied:
If an exception comes out you know that there was a problem creating the folder without even looking at the code

fernandotorch commented on algorithms/a3c/main.py:
```diff
@@ -40,80 +54,276 @@
-    counter = mp.Value('i', 0)
-    lock = mp.Lock()
+    # Checkpoint creation/loading below
+    checkpoint_counter = False
```
Checkpoint_counter is only used in the else, so create it within it not here

beduffy replied:
Nope, it's used after the else

fernandotorch commented on algorithms/a3c/main.py:
```diff
@@ -40,80 +54,276 @@
+              'made here: {}'.format(args.experiment_path, args.checkpoint_path))
+        os.makedirs(args.checkpoint_path)
+    else:
+        print('Checkpoints path already exists at path: {}'.format(args.checkpoint_path))
```
```suggestion
        print('Using existing checkpoints at path: {}'.format(args.checkpoint_path))
```

fernandotorch commented on algorithms/a3c/main.py:
```diff
@@ -40,80 +54,276 @@
+                print("=> loaded checkpoint '{}' (total_length {})"
+                      .format(checkpoint_to_load, checkpoint['total_length']))
+        else:
+            print('No model checkpoint to load')
```
```suggestion
            print('No model checkpoint found! Please make sure that you saved the file in such and such structure)
```

fernandotorch commented on algorithms/a3c/main.py:
```diff
@@ -40,80 +54,276 @@
+        if checkpoint_paths:
+            # Take checkpoint path with most experience
+            # e.g. 2000 from checkpoint_total_length_2000.pth.tar
+            checkpoint_file_name_ints = [
```
make this more compact and clear by storing only the max:
```suggestion
            idx_of_latest = 0
            for x in checkpoint_paths:
                steps_in_x = int(x.split('/')[-1].split('.pth.tar')[0].split('_')[-1])
                idx_of_latest = max(steps_in_x, idx_of_latest)
```

fernandotorch replied:
I wonder if there is a way to simply get the last modified file. Otherwise I stick to my suggestion above

fernandotorch commented on algorithms/a3c/main.py:
```diff
@@ -40,80 +54,276 @@
+                for x in checkpoint_paths]
+            idx_of_latest = checkpoint_file_name_ints.index(max(checkpoint_file_name_ints))
+            checkpoint_to_load = checkpoint_paths[idx_of_latest]
+            print('Attempting to load latest checkpoint: {}'.format(checkpoint_to_load))
```
```suggestion
            print('Loading latest checkpoint: {}'.format(checkpoint_to_load))
```

fernandotorch commented on algorithms/a3c/main.py:
```diff
@@ -40,80 +54,276 @@
+            print('Attempting to load latest checkpoint: {}'.format(checkpoint_to_load))
+
+            # load checkpoint and unpack values
+            if os.path.isfile(checkpoint_to_load):
```
you could use os.walk instead of glob.glob to extract only files and not need to check it here

fernandotorch commented on algorithms/a3c/main.py:
```diff
@@ -40,80 +54,276 @@
+
+            # load checkpoint and unpack values
+            if os.path.isfile(checkpoint_to_load):
+                print("Successfully loaded checkpoint {}".format(checkpoint_to_load))
```
Before this print you didn't check if the checkpoint has valid data or anything. I would remove it completely or make a try except to catch `KeyError` exceptions

fernandotorch commented on algorithms/a3c/main.py:
```diff
@@ -40,80 +54,276 @@
+                args.num_backprops = checkpoint.get('num_backprops', 0)
+                checkpoint_counter = checkpoint.get('counter', False)  # if not set, set to 0 later
+
+                print('Values from checkpoint: total_length: {}. episode_number: {}'.format(
```
```suggestion
                print('Checkpoint loaded with total_length: {} and episode_number: {}'.format(
```

fernandotorch commented on algorithms/a3c/main.py:
```diff
@@ -40,80 +54,276 @@
+    counter = mp.Value('i', 0 if not checkpoint_counter else checkpoint_counter)
+    lock = mp.Lock()
+    try:
+        if not args.synchronous:
```
pep8: if args.synchronous what you put in the else, else what you put in this if. Always formulate positive ifs if you can

fernandotorch commented on algorithms/a3c/model.py:
```diff
@@ -12,22 +13,45 @@
     42x42 -> (42 − 3 + 2)÷ 2 + 1 = 21x21 after 1 layer
     11x11 after 2 layers -> 6x6 after 3 -> and finally 3x3 after 4 layers
     Therefore lstm input size after flattening would be (3 * 3 * num_filters)
     """
```
Add: ```We assume that the same kernel_size, padding and stride is used in all convolutional layers```

fernandotorch commented on algorithms/a3c/model.py:
```diff
@@ -12,22 +13,45 @@
+
+    return width * height * num_filters
+
+def calculate_input_width_height_for_A3C_LSTM_GA(resolution):
```
Two spaces between functions? We need to be consistent and either use 1 or 2 always (recommended 2 from my side)

beduffy replied:
let's have a conversation but i might be willing to do 2

beduffy replied:
class functions can be separated by 1 according to my PyCharm?

fernandotorch commented on algorithms/a3c/model.py:
```diff
@@ -12,22 +13,45 @@
+
+def calculate_input_width_height_for_A3C_LSTM_GA(resolution):
+    """
+    Does't assume square resolution image (since ViZDoom default is not square).
```
The previous function doesn't assume square resolution anymore so I think we can remove this line from the docstring

fernandotorch commented on algorithms/a3c/model.py:
```diff
@@ -12,22 +13,45 @@
+    Does't assume square resolution image (since ViZDoom default is not square).
+    Similar to the calculate_lstm_input_size_for_A3C function except that there are only
+    3 conv layers and there is variation among the kernel_size, stride, the number of channels
+    and there is no padding. Therefore these are hardcoded. Check A3C_LSTM_GA class for these numbers.
```
Line with `numbers` is too long

fernandotorch commented on algorithms/a3c/model.py:
```diff
@@ -12,22 +13,45 @@
+    Similar to the calculate_lstm_input_size_for_A3C function except that there are only
+    3 conv layers and there is variation among the kernel_size, stride, the number of channels
+    and there is no padding. Therefore these are hardcoded. Check A3C_LSTM_GA class for these numbers.
+    Also, returns tuple representing (width, height, num_output_filters) instead of size
```
As in its equivalent for A3C, it returns a tuple bla

beduffy replied:
A3C's one returns the size not the tuple

fernandotorch commented on algorithms/a3c/model.py:
```diff
@@ -12,22 +13,45 @@
+
+    return width, height, 64
 
 def normalized_columns_initializer(weights, std=1.0):
```
Why do I need column normalization? Add an explanation here

fernandotorch replied:
Also ellaborate more on this particular type of normalization being used as opposed to other ones

fernandotorch commented on algorithms/a3c/model.py:
```diff
@@ -36,10 +60,9 @@ def normalized_columns_initializer(weights, std=1.0):
-
 def weights_init(m):
     classname = m.__class__.__name__
     if classname.find('Conv') != -1:
```
I see that you are initializing layers differently, but what are the magic numbers like 6? Can you make a docstring for the function?

fernandotorch commented on algorithms/a3c/model.py:
```diff
@@ -60,26 +83,26 @@ def weights_init(m):
         self.conv3 = nn.Conv2d(32, 32, 3, stride=2, padding=1)
         self.conv4 = nn.Conv2d(32, 32, 3, stride=2, padding=1)
 
         # assumes square image
```
Not anymore
```suggestion
    
```

fernandotorch commented on algorithms/a3c/model.py:
```diff
@@ -111,3 +132,94 @@ def forward(self, inputs):
+    could learn and language ground itself in the meaning of "red" with a filter that learns this
+    mapping. There is also a time embedding layer to help stabilize value prediction.
+    Only 3 conv layers compared to ActorCritic's 4 layers.
+    Originally ran on ViZDoom environment with a non-square image state
```
non-square is irrelevant. Remove that

fernandotorch commented on algorithms/a3c/model.py:
```diff
@@ -111,3 +132,94 @@ def forward(self, inputs):
+
+class A3C_LSTM_GA(torch.nn.Module):
+    """
+    Very similar to the above ActorCritic but has Gated Attention (GA) and processes an instruction
```
I know you love vertical space, but I would split this at least into two paragraphs to help the reader. You are also saying "grounding" but not giving a clear explanation of why this is grounding or what it is.

beduffy replied:
replaced with:

"Very similar to the above ActorCritic but has Gated Attention (GA) and processes an instruction
    which is a part of the input state using a GRU. There is also a time embedding layer to help 
    stabilize value prediction and only 3 conv layers compared to ActorCritic's 4 layers.
    Originally ran on the ViZDoom environment. The above ASCII art figures and commented code 
    below should make the flow quite clear.    
    
    The gated attention multi-modal fusion module enables the policy to focus on certain parts
    of the input image given the instruction e.g. for the instruction "Go to the red cup", a 
    specific attention filter could be learned which would enable the agent to language ground 
    itself in the meaning of both "red", "cup" and "go to". Language grounding is the ability to 
    map meaning within symbols+language into real world objects and goals. "

fernandotorch commented on algorithms/a3c/model.py:
```diff
@@ -111,3 +132,94 @@ def forward(self, inputs):
+
+
+class A3C_LSTM_GA(torch.nn.Module):
+    """
```
It would be nice to add an ASCII art diagram of the architecture. Something like:
```
     ___________
cx--|   LSTM    | --- Critic Linear -- Q
hx--|___________|
```
Obviously my example is wrong, but just an idea

fernandotorch replied:
Also applicable to A3C above

beduffy replied:
hahaha, might be worth it but it might take me 15 minutes haha

fernandotorch commented on algorithms/a3c/model.py:
```diff
@@ -111,3 +132,94 @@ def forward(self, inputs):
+            self.num_output_filters = calculate_input_width_height_for_A3C_LSTM_GA(resolution)
+        self.lstm_cell_size = self.output_width * self.output_height * self.num_output_filters
+
+        # Instruction Processing
```
```suggestion
        # Language Instruction Processing
```

beduffy replied:
Used:
`# Natural Language Instruction Processing`
instead

fernandotorch commented on algorithms/a3c/model.py:
```diff
@@ -111,3 +132,94 @@ def forward(self, inputs):
+        # Gated-Attention layers
+        self.attn_linear = nn.Linear(self.gru_hidden_size, 64)
+
+        # Time embedding layer, helps in stabilizing value prediction
```
Could you explain more to me about the time embedding layer and why it stabilizes value?

beduffy replied:
I don't actually know "exactly" how but let's work it out here:

if episode is 1000 steps long and the the step number from 0 to 999 get's passed in as tx, and extra self.time_emb_dim dimensions are input to critic+actor to help use time to your advantage (e.g. rush when you're running out of time) or purely just to know that the min reward you can get in 1000 steps with movement reward  -0.001 is -1.0). The agent can work out in sparse reward settings therefore that if there are only 3 steps left it's value prediction will be accurate towards the end and no cup in sight?

How about something like that?

beduffy replied:
i don't see exactly how it can help the actor though? Should you act drastically differently with 30 steps left than with 300? Maybe if you have to choose between going to the closer cup or not

beduffy replied:
Yusun helped me figure this out. Value prediction isn't given time input so for the same image state, if there is 1 step left in episode or 900, the value head can't learn to predict less value for the 1st case. Added this comment:

```
# Time embedding layer, helps in stabilizing value prediction e.g. if only 1 step is left
# in episode not much reward can be earned vs the same image state with 500 steps left
```

fernandotorch commented on algorithms/a3c/model.py:
```diff
@@ -111,3 +132,94 @@ def forward(self, inputs):
+        x_attention = x_attention.unsqueeze(2).unsqueeze(3)
+        x_attention = x_attention.expand(1, self.num_output_filters, self.output_height,
+                                         self.output_width)
+        assert x_image_rep.size() == x_attention.size()
```
No asserts. Check and raise if it doesn't work

beduffy replied:
hmmmm, i really like the conciseness of this assert and i can add a comment?
check and raise is lame to have in a forward function? 2 lines and then you have ugly raise ValueError's in your beauitfully clear forward function

beduffy committed:
Added EPIC 3 figures of ASCII art for A3C_LSTM_GA architecture and other comments/todos in model.py, my_optim.py and train.py

Removed square image talk.
In README.md mentioned needing to install ViZDoom for ViZDoom command
2 lines between functions

beduffy committed:
Added ActorCritic ASCII chart, cleaned up comments in model.py and made clearer. Split A3C_LSTM_GA docstring in two and explained langauge grounding. self.num_output_filters = 64. Explained GA expand notation and depth slices

 Removed learning during training from README.md

beduffy committed:
Added 13 todos to main.py, fixed big bug on storing all_rewards_in_episode only in done, put 4 appends before done (why was it otherwise in the first place?) and removed 2nd "if done", fixed ai2thor_examples.py multithreading example fps printing

removed printing of 3d euclidean

fernandotorch commented on algorithms/a3c/test.py:
```diff
@@ -9,36 +9,49 @@
     env.seed(args.seed + rank)
 
-    model = ActorCritic(env.observation_space.shape[0], env.action_space.n, args.frame_dim)
+    if env.task.task_has_language_instructions:
```
Could we have a convention of tasks with language starting with "Language" or sth like that? so we would check `if env.task.__name__.startswith("Language"):` That would look better IMO

beduffy replied:
I do have `NaturalLanguage` at the start of all of them and `NL` at the start of the config files. But a boolean is just easier and clearer no??
Booleans are MUCH clearer than startswith stuff which could be wrong. That way requires us to stick to the standard.

fernandotorch replied:
and isn't sticking to a standard a good thing? Rather than having to add 1 extra param

beduffy replied:
it's added automatically in the NaturalLanguageBaseTask? 
0 work, super clear, very readable, very DRY and doesn't require us to forcibly stick to a standard as we create 100+ tasks

beduffy committed:
Put input_size functions as static methods for A3C and A3C_LSTM_GA, Added eid info to README, default reward is 1 for 5 objects

removed printing of num ranbdom actions at init

beduffy committed:
Added show_instance_segmentation (binary mask matplotlib image on specified object types) and example to pdb_interactive py file

beduffy committed:
Added _process_frame42 as staticmethod in env_atari.py. In env_vizdoom code: No more import star for vizdoom. Style changes, docs and avoid repetition

beduffy committed:
Removed seed init param from ai2thor_env, added lastObjectClosed, lastObjectOpened, lastObjectPickedUp, lastObjectPut, lastObjectPutReceptacle. (need to test). Added todos, time embedding explanation

fernandotorch commented on gym_ai2thor/tasks.py:
```diff
@@ -2,8 +2,10 @@
+from gym_ai2thor.task_utils import get_word_to_idx, check_if_focus_and_close_enough_to_object_type
 
 
 class TaskFactory:
```
let's get rid of this task entirely by calling `getattr(gym_ai2thor.tasks, config['task']['task_name'])(**config)` from AI2ThorEnv

fernandotorch commented on gym_ai2thor/tasks.py:
```diff
@@ -93,10 +112,226 @@ def transition_reward(self, state):
+        return super(NaturalLanguageNavigateToObjectTask, self).reset()
+
+
+class NaturalLanguagePickUpObjectTask(NaturalLanguageBaseTask):
```
I would restructure this to have this class inherit from both NaturalLanguageBaseTask and PickupTask. We can discuss this in slack

fernandotorch commented on algorithms/a3c/test.py:
```diff
@@ -9,36 +9,49 @@
     env.seed(args.seed + rank)
 
-    model = ActorCritic(env.observation_space.shape[0], env.action_space.n, args.frame_dim)
+    if env.task.task_has_language_instructions:
```
Totally different issue. What happens if the task is not natural language? task.has_language_instructions will raise a ValueError

fernandotorch commented on gym_ai2thor/envs/ai2thor_env.py:
```diff
@@ -66,106 +73,185 @@ def __init__(self, seed=None, config_file='config_files/config_example.json', co
+                            obj['objectType'] in self.allowed_objects['openables']:
                         closest_openable = obj
                         distance = closest_openable['distance']
                     if closest_openable:
```
This is wrong. If there are two objects to close and the furthest one is first in visible_objects iterable you will close both the furthest and the closest. Moreover, we are executing the close action as many times as remaining objects in visible_objects after the closest_openable. Unindenting this if block will solve the issue. The same applies to "CloseObject" action

fernandotorch commented on gym_ai2thor/envs/ai2thor_env.py:
```diff
@@ -66,106 +73,185 @@ def __init__(self, seed=None, config_file='config_files/config_example.json', co
                 closest_openable = None
                 for obj in visible_objects:
                     # look for closest closed receptacle to open it
                     if obj['openable'] and obj['distance'] < distance and not obj['isopen'] and \
```
I'd like to make this more understandable. Can we take the whole condition in the if to a variable called `is_closest_receptacle` and then just do `if is_closest_receptacle`? Then the comment above this line won't be needed anymore

fernandotorch replied:
Analogously we can do the same for `is_closest_pickupable` and `is_closest_opened_receptacle` if you know what I mean

fernandotorch commented on gym_ai2thor/envs/ai2thor_env.py:
```diff
@@ -27,35 +31,38 @@
+    This class wraps the ai2thor environment
     """
-    def __init__(self, seed=None, config_file='config_files/config_example.json', config_dict=None):
+    def __init__(self, config_file='config_files/default_config.json', config_dict=None):
```
there is no seed parameter although it is used in main.py?

fernandotorch commented on gym_ai2thor/envs/ai2thor_env.py:
```diff
@@ -66,106 +73,185 @@ def __init__(self, seed=None, config_file='config_files/config_example.json', co
         if not self.config['pickup_put_interaction']:
             self.action_names = tuple([action_name for action_name in self.action_names if 'Pickup'
                                        not in action_name and 'Put' not in action_name])
+        if self.config.get('put_interaction'):
```
Can we simplify all these gets in one single for loop? 
```self.action_names = list(ALL_POSSIBLE_ACTIONS.copy())
for action_name in self.action_names:
    if action_name.startswith('Look') and not self.config.get('lookupdown_actions, True) or
    if action_name.startswith('Move') and not self.config.get('moveupdown_actions, True):
        self.action_names.remove(acton_name)
self.action_names = tuple(self.action_names)
```
And so on. It might not be 100% correct but you get the idea

beduffy committed:
Fixed if indent on open and close objects, simplified logic into boolean vars so if statement looks easy to read, set last vars to None

beduffy committed:
Renamed to has_language_instructions, todos, notes

beduffy committed:
Added more todos, lots of unfinished stuff (pushing for remote work purposes), writer problems, printing latest_sys_args, args.env, 3 objects config
